### ISSUE TITLE
When a resource with multiple templates select a different template, then there are not orphan objects

### ISSUE THREAD

**Issue by zrob** [vmware-tanzu/cartographer#592]
As a follow on to #573 

```
Given: A supply chain that has two templates for a resource, templateA and templateB
AND: a workload that results in templateA being selected
When: A workload is modified to result in templateB being selected
Then: The objects created by templateA are cleaned up
```

This can resolve both ytt templates and options that generate orphans.

**Comment by cirocosta** [comment]
some thoughts from IPM on a possible way of addressing all cases - with pseudo-code, I imagine something in the lines of the following should do it 

```go
// (on reconciler run)

// (1)
stampedObjects := realizer.Realize()

// (2)
directChildren := repository.FindChildren(workload)

orphans := diff(stampedObjects, directChildren)
garbageCollect(orphans)
```

but, in practice that can be a little tricky:

1. `stampedObjects` does not _always_ include all the objects that represents the current set of objects that should be stamped - e.g., if we have a supplychain like (`a -> b -> c`), and we have had successful creation for all 3 objects (a, b, and c), but for some reason (e.g., template for `a` is messed up), then `a` will prevent the other resources from being updated and thus `stampedObjects` will be empty (well, _perhaps_ this is something that should be changed anyway? or ... perhaps that's a reconciliation step that shouldn't involve `gc` anyway as something wrong happened and we should try later)

2. `directChildren`: finding those by just looking at the cluster by searching across all GVRs (I imagine that's the only way?) could get really expensive in a cluster with many custom resources (perhaps there's an optimization here? e.g., list very rarely but somehow leverage repository's cache? not sure ..)

**Comment by emmjohnson** [comment]
It doesn't appear that we can "list all objects by label" - https://stackoverflow.com/questions/54675382/how-to-list-all-kubernetes-objects-with-specific-label-using-client-go

**Comment by emmjohnson** [comment]
Spike using configmaps - https://github.com/vmware-tanzu/cartographer/compare/spike/592-orphans?expand=1

**Comment by emmjohnson** [comment]
Other ideas:
- using a db
- list all objects by getting all api groups
  - some sort of interval with list all objects

**Comment by emmjohnson** [comment]
Blocked on - https://github.com/vmware-tanzu/cartographer/pull/519#issuecomment-1028020489

**Comment by cirocosta** [comment]
> _Blocked on - https://github.com/vmware-tanzu/cartographer/pull/519#issuecomment-1028020489_

now, https://github.com/vmware-tanzu/cartographer/pull/634

note that in the RFC there's this point made

> _If there is significant user concern, a toggle could be developed to partially or fully disable tracing. Such a toggle is not in scope for this RFC._

that sounds _very_ relevant to this issue if we go with the approach on relying on `status.resources` - if we envision a toggle for such functionality, then I imagine we'd end up having to have either a partial gc behavior / no gc at all?
